@xen-orchestra/xapi/vdi.mjs

```diff
@@ -95,96 +96,114 @@ class Vdi {
-            info('found changed blocks', changedBlocks)
-          } catch (error) {
-            // do not fail if CBT is not enabled/working
-            info('no changed block', error)
-          }
-        }
-        nbdClient = await this._getNbdClient(ref, { nbdConcurrency })
+
+    let nbdClient, // optional : the nbd client used to transfer this vdi
+      exportStream, // the stream read from the XAPI
+      stream, // the stream that will be returned (exportStream or using NBD)
+      taskRef, // the reference to the export stream (if created manually)
+      changedBlocks, // the CBT list of blocks
+      baseParentUuid // the uuid of the parent used as a base for a delta export
+
+    if (baseRef !== undefined && preferNbd && cbt_enabled) {
+      // use CBT if possible
+      // call to list changed blocks must be done before the vdi is mounted for NBD export
+      try {
+        changedBlocks = await this.VDI_listChangedBlock(ref, baseRef)
+        baseParentUuid = await this.getField('VDI', baseRef, 'sm_config').then(sm_config => sm_config?.['vhd-parent'])
+
+        info('found changed blocks', changedBlocks)
+      } catch (error) {
+        // do not fail if CBT is not enabled/working
+        info('no changed block', error)
       }
-      // the raw nbd export does not need to peek ath the vhd source
-      if (nbdClient !== undefined && format === VDI_FORMAT_RAW) {
-        stream = createNbdRawStream(nbdClient)
-      } else {
-        if (changedBlocks === undefined) {
-          // raw export without nbd or vhd exports needs a resource stream
-          // also metadata vdis (vdi that have a data_destroy) can't export a stream
-          stream = exportStream = (
-            await this.getResource(cancelToken, '/export_raw_vdi/', {
-              query,
-              task: await this.task_create(`Exporting content of VDI ${vdiName}`),
-            })
-          ).body
-        }
+    }
 
-        if (nbdClient !== undefined && format === VDI_FORMAT_VHD) {
-          const taskRef = await this.task_create(
-            `Exporting content of VDI ${vdiName} using NBD ${changedBlocks !== undefined ? ' and CBT' : ''}`
-          )
-          exportStream = stream
-          stream = await createNbdVhdStream(nbdClient, exportStream, {
-            changedBlocks,
-            vdiInfos: { size, uuid, parentUuid: baseParentUuid },
-          })
-          stream.on('progress', progress => this.call('task.set_progress', taskRef, progress))
-          finished(stream, () => {
-            nbdClient.disconnect()
-            exportStream?.destroy() // ensure the source stream is really closed
-
-            this.task_destroy(taskRef).catch(warn)
-          })
+    // really connect to NBD server
+    if (preferNbd) {
+      nbdClient = await this._getNbdClient(ref, { nbdConcurrency })
+      // disconnect on failure or when transfer is finished
+      $defer.onFailure(() => nbdClient?.disconnect())
```

Comment:
When can `nbdClient` be undefined?

Reply:
if the connexion fails / no NBD network / no path to te vdi through NBD

Reply:
I see.

It was not obvious to me that `_getNbdClient` could return `undefined`, maybe a name like `_tryGetNbdClient` would be clearer?

In that case, you can make the test immediate and explicit:

```suggestion
      if (nbdClient !== undefined) {
        $defer.onFailure(() => nbdClient.disconnect())
      }
```

Reply:
I think the ? is exactly here for this . ok for the renaming

Reply:
I agree it will work equally, but, to me, it means a different things semantically:
- the `if`: only disconnect at the end if the previous call returned an NBD client
- the `?`: the NBD client may have been removed when we reach this code